### ISSUE TITLE
control_dependencies of ExponentialMovingAverage in cifar10_multi_gpu_train.py

### ISSUE THREAD

**Issue by c149028** [tensorflow/tensorflow#2131]
There are two ExponentialMovingAverage in cifar10_multi_gpu_train.py.  

- For model parameters https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/image/cifar10/cifar10_multi_gpu_train.py#L220  
Should the `apply` operation of ExponentialMovingAverage be called after the update of parameters? 

```python 
with tf.control_dependencies([apply_gradient_op]):
    train_op = variable_averages.apply(tf.trainable_variables())
```

- For loss
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/image/cifar10/cifar10_multi_gpu_train.py#L105-L106  
Are these two lines redundant? Loss value won't change within a single run.

P.S. The document of ExponentialMovingAverage (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/moving_averages.py#L172-L177) says

> ``` python 
maintain_averages_op = ema.apply([var0, var1])
  # Create an op that will update the moving averages after each training
  # step.  This is what we will use in place of the usual training op.
with tf.control_dependencies([opt_op]):
    training_op = tf.group(maintain_averages_op)```

If we want to update the moving averages after the training step, should we call `ema.apply` within the context of tf.control_dependencies([opt_op])?

**Comment by sherrym** [comment]
1. By doing the following:
 with tf.control_dependencies([apply_gradient_op]):
    train_op = variable_averages.apply(tf.trainable_variables())

The parameters are updated before the variable_averages.

2. This is multi-gpu run. We are trying to return a snapshot of total_loss in case it's being updated by the other thread.

3. I don't fully understand your question. Are you asking whether the implementation you quoted is correct, or you are trying to do something else different, and wonder how you should do it?

Thanks,
Sherry

**Comment by c149028** [comment]
Hi Sherry,

Suppose there is a parameter `p` initialized by 0. During training time the value of `p` are  
0 -> 1 -> 2 -> 3 -> 4  
Is it possible that the result of `ExponentialMovingAverage` become the moving average of 0, 1, 1, 3, 4 when we didn't specify the dependencies between `variable_averages.apply` and `apply_gradient_op`? 

-----

I cannot figure out under what circumstances will `total_loss` be modified by the other thread within a single `sess.run`.
Why we need
``` python
with tf.control_dependencies([loss_averages_op]):
    total_loss = tf.identity(total_loss)
```
to ensure `loss_averages_op` be executed before using `total_loss`?

**Comment by girving** [comment]
This fits better as a question on StackOverflow, since it doesn't seem to be about a fixable issue in TensorFlow.
